lighthouse-core/lib/dependency-graph/base-node.js

```diff
@@ -168,96 +168,79 @@ class BaseNode {
 
   /**
    * Clones the entire graph connected to this node filtered by the optional predicate. If a node is
-   * included by the predicate, all nodes along the paths between the two will be included. If the
-   * node that was called clone is not included in the resulting filtered graph, the method will throw.
+   * included by the predicate, all nodes along the paths between the node and the root will be included. If the
+   * node this was called on is not included in the resulting filtered graph, the method will throw.
    * @param {function(Node):boolean} [predicate]
    * @return {Node}
    */
   cloneWithRelationships(predicate) {
     const rootNode = this.getRootNode();
 
-    /** @type {function(Node): boolean} */
-    let shouldIncludeNode = () => true;
-    if (predicate) {
-      const idsToInclude = new Set();
-      rootNode.traverse(node => {
-        if (predicate(node)) {
-          node.traverse(
-            node => idsToInclude.add(node.id),
-            node => node._dependencies.filter(parent => !idsToInclude.has(parent))
-          );
-        }
-      });
-
-      shouldIncludeNode = node => idsToInclude.has(node.id);
-    }
+    /** @type {Map<string, Node>} */
+    const idsToIncludedClones = new Map();
 
-    const idToNodeMap = new Map();
-    rootNode.traverse(originalNode => {
-      if (!shouldIncludeNode(originalNode)) return;
-      const clonedNode = originalNode.cloneWithoutRelationships();
-      idToNodeMap.set(clonedNode.id, clonedNode);
+    // Walk down dependents.
+    rootNode.traverse(node => {
+      // Find new and includable nodes.
+      if (idsToIncludedClones.has(node.id)) return;
+      if (predicate && !predicate(node)) return;
+
+      // Walk back up dependencies, cloning all nodes in the path to this one.
+      node.traverse(
+        node => idsToIncludedClones.set(node.id, node.cloneWithoutRelationships()),
+        node => node._dependencies.filter(parent => !idsToIncludedClones.has(parent.id))
```

Comment:
can we get a comment on this one that explains the filter while we're at it? :) it even took me a sec to remember why this was happening at all

Reply:
> can we get a comment on this one that explains the filter while we're at it? 

good call